### PR TITLE
Implementation of Decorated Cospans

### PR DESCRIPTION
**Pull request by mehalter**
As I am working on [AlgebraicPetri.jl](https://github.com/AlgebraicJulia/AlgebraicPetri.jl), I am putting together an implementation of decorated cospans in Catlab to help work with open systems.

### CODE REVIEW

**Comment by mehalter** [comment]
I added some abstract types for Functor, Laxitor, and LaxMonoidalFunctor for more general use in the future. Let me know if you want to move this to a different place or not include it in there.

**Comment by mehalter** [comment]
I finished the implementation of AlgebraicPetri and this is the stuff I added to get Decorated Cospans working with Catlab. Let me know if any of this should change before merging

File: src/categorical_algebra/ShapeDiagrams.jl [hunk 1]
```diff
@@ -1,10 +1,20 @@
 """ Diagrams of a given shape.
 """
 module ShapeDiagrams
-export Span, Cospan, left, right, apex, base
+export AbstractFunctor, AbstractLaxitor, LaxMonoidalFunctor,
```
Comment (epatters) [inline src/categorical_algebra/ShapeDiagrams.jl:4]: `AbstractLaxitor` should be spelled `AbstractLaxator`.

File: src/categorical_algebra/ShapeDiagrams.jl [hunk 2]
```diff
@@ -1,10 +1,20 @@
...
+       Span, Cospan, DecoratedCospan, left, right, apex, base,
+       decorator, decoration, undecorate
 
 using ...Theories: dom, codom
 
+abstract type AbstractFunctor end
```
Comment (epatters) [inline src/categorical_algebra/ShapeDiagrams.jl:10]: Strictly speaking, stuff about lax monoidal functors doesn't belong in this module, but I think that's OK for now. Once we have a fully generic system for decorated cospans, we can create a separate module for decorated cospans and put everything there.

Comment (mehalter) [reply src/categorical_algebra/ShapeDiagrams.jl:10]: Yeah I agree it doesn't really fit in this module, so we can definitely move it once it becomes sufficiently sophisticated.

File: src/categorical_algebra/ShapeDiagrams.jl (hunk 1 above)
Comment (mehalter) [reply src/categorical_algebra/ShapeDiagrams.jl]: Fixed in latest commit.

**Pull request merged**